src/main/java/networkbook/model/person/Course.java

```diff
@@ -3,43 +3,120 @@
+            + "3. Course number must be 4 numbers long\n"
+            + "4. Variant code (optional) is a single letter\n"
+            + "5. Start date and End date are optional and must follow the format DD-MM-YYYY";
+    public static final String DATE_CONSTRAINTS = "End date should occur after the start date.";
 
     /*
-     * The first character of the course must not be a whitespace,
-     * otherwise " " (a blank string) becomes a valid input.
+     * The first 2-3 characters of a course must be uppercase.
+     * The following 4 characters must be numbered.
+     * The course may or may not end with a uppercase variant letter.
      */
-    public static final String VALIDATION_REGEX = "[^\\s].*";
+    public static final String VALIDATION_REGEX = "^[A-Z]{2,3}[0-9]{4}[A-Z]?$";
 
-    public final String value;
+    public final String courseCode;
+    private Optional<String> startDate;
+    private Optional<String> endDate;
 
     /**
      * Constructs an {@code Course}.
      *
-     * @param course A valid course.
+     * @param course A valid course with no start or end date.
      */
     public Course(String course) {
         requireNonNull(course);
         checkArgument(isValidCourse(course), MESSAGE_CONSTRAINTS);
-        value = course;
+        courseCode = course;
+        startDate = Optional.empty();
+        endDate = Optional.empty();
+    }
+
+    /**
+     * Constructs an {@code Course} with start and/or end dates.
+     *
+     * @param course A valid course name.
+     * @param dates An array containing the start date and possibly end date of taking the course.
+     */
+    public Course(String course, String[] dates) {
+        requireNonNull(course);
+        checkArgument(isValidCourse(course), MESSAGE_CONSTRAINTS);
+        courseCode = course;
+        assert (dates.length > 0 && dates.length <= 2);
+        startDate = Optional.empty();
+        endDate = Optional.empty();
+        for (String date : dates) {
+            requireNonNull(date);
+            checkArgument(isValidDate(date), MESSAGE_CONSTRAINTS);
+            if (startDate.isEmpty()) {
+                startDate = Optional.ofNullable(date);
+            } else {
+                endDate = Optional.ofNullable(date);
+            }
+        }
+        if (!endDate.isEmpty()) {
+            checkArgument(areChronologicalDates(startDate.get(), endDate.get()), DATE_CONSTRAINTS);;
+        }
     }
 
     /**
      * Returns true if a given string is a valid course.
      */
-    public static boolean isValidCourse(String test) {
-        return test.matches(VALIDATION_REGEX);
+    public static boolean isValidCourse(String course) {
+        return course.matches(VALIDATION_REGEX);
+    }
+
+    /**
+     * Returns true if a given string is a valid date.
+     * For defensive purposes the format of the date will be limited to DD-MM-YYYY format.
+     */
+    public static boolean isValidDate(String date) {
```

Comment:
Can consider moving this method into a util class (maybe after v1.2) as it seems like it could be applicable outside of Course